keras_list_mapper/keras_list_mapper.py

```diff
@@ -211,16 +217,65 @@ def call(self, inputs, *args, **kwargs):  # pylint: disable=unused-argument
+            layer_output = self.layer(inner_inputs, *args, **kwargs)
             if self.state_shape is not None:
-                state, output = y
+                state, *layer_output = layer_output
                 states = tf.tensor_scatter_nd_update(states, indices, state)
-            else:
-                output = y
+                if len(layer_output) == 1:
+                    layer_output = layer_output[0]
 
             ix = tf.add(tf.gather_nd(rt.row_splits, indices), xcol)
             ix = tf.expand_dims(ix, 1)
 
-            outputs = tf.tensor_scatter_nd_update(outputs, ix, output)
+            outputs = nest.map_structure(
+                lambda x, y: tf.tensor_scatter_nd_update(x, ix, y),
+                outputs, layer_output)
+
+        return nest.map_structure(
+            lambda x: tf.RaggedTensor.from_row_splits(x, rt.row_splits),
+            outputs)
+
+    @staticmethod
+    def _remove_timesteps(dims: tf.TensorShape, batch_input: bool
+                          ) -> tf.TensorShape:
+        if batch_input:
+            return dims
+        dims = dims.as_list()
+        return tf.TensorShape((dims[0], *dims[2:]))
+
+    def _batch_inputs(self, input_shapes) -> Union[bool, List[bool]]:
+        batch_inputs = 0 in self.batch_inputs
+        if nest.is_sequence(input_shapes):
+            # Check if the input was declared as batch input or if the second
+            # dim is not ragged.
+            batch_inputs = [i in self.batch_inputs or shape[1] is not None
+                            for i, shape in enumerate(input_shapes)]
+        return batch_inputs
+
+    def compute_output_shape(self, input_shape):
+        input_shape = tf_utils.convert_shapes(input_shape, to_tuples=False)
+
+        batch_inputs = self._batch_inputs(input_shape)
+        child_input_shape = nest.map_structure(self._remove_timesteps,
+                                               input_shape,
+                                               batch_inputs)
+        if self.state_shape is not None:
+            batch_size = nest.flatten(tf_utils.convert_shapes(input_shape))[0]
+            state_shape = tf.TensorShape((batch_size, *self.state_shape))
+            child_input_shape = [state_shape, *nest.flatten(child_input_shape)]
+
+        child_output_shape = self.layer.compute_output_shape(child_input_shape)
 
-        return tf.RaggedTensor.from_row_splits(outputs, rt.row_splits)
+        if self.state_shape is not None:
+            child_output_shape = child_output_shape[1:]
+            if len(child_output_shape) == 1:
+                child_output_shape = child_output_shape[0]
+        child_output_shape = tf_utils.convert_shapes(
+            child_output_shape, to_tuples=False)
+        timesteps = nest.flatten(tf_utils.convert_shapes(input_shape))[1]
+
+        def insert_timesteps(dims: tf.TensorShape):
+            dims = dims.as_list()
+            return tf.TensorShape([dims[0], timesteps] + dims[1:])
```

Comment:
The ListMapper returns a ragged Tensor. There is a new layer API compute_output_signature that is suppose to be able to convey this information. Would it makes sense to use that instead ?

Reply:
Even though `compute_output_signature` seems to be called only for [dynamic layers](https://github.com/tensorflow/tensorflow/blob/ec5ff5df834bcd32941ffac649950558f2adea1f/tensorflow/python/keras/engine/base_layer.py#L864), it is indeed useful internally.
It is now implemented and used to create the output placeholders, simplifying the code in `__call__`.

Reply:
Even though `compute_output_signature` seems to be called only for [dynamic layers](https://github.com/tensorflow/tensorflow/blob/ec5ff5df834bcd32941ffac649950558f2adea1f/tensorflow/python/keras/engine/base_layer.py#L864), it is indeed useful internally.
It is now implemented and used to create the output placeholders, simplifying the code in `__call__`.